rtos/ConditionVariable.h

```diff
@@ -0,0 +1,186 @@
+ * called and there are one or more waiters and one or more threads attempting
+ * to acquire the condition variable's mutex the order in which the mutex is
+ * acquired is undefined.
+ * * The behavior of ConditionVariable::wait and ConditionVariable::wait_for
+ * is undefined if the condition variable's mutex is locked more than once by
+ * the calling thread.
+ * * Spurious notifications (not triggered by the application) can occur
+ * and it is not defined when these occur.
+ *
+ * @note Synchronization level: Thread safe
+ *
+ * Example:
+ * @code
+ * #include "mbed.h"
+ *
+ * Mutex mutex;
+ * ConditionVariable cond(mutex);
+ *
+ * // These variables are protected by locking mutex
+ * uint32_t count = 0;
+ * bool done = false;
+ *
+ * void worker_thread()
+ * {
+ *     mutex.lock();
+ *     do {
+ *         printf("Worker: Count %lu\r\n", count);
+ *
+ *         // Wait for a condition to change
+ *         cond.wait();
+ *
+ *     } while (!done);
+ *     printf("Worker: Exiting\r\n");
+ *     mutex.unlock();
+ * }
+ *
+ * int main() {
+ *     Thread thread;
+ *     thread.start(worker_thread);
+ *
+ *     for (int i = 0; i < 5; i++) {
+ *
+ *         mutex.lock();
+ *         // Change count and notify waiters of this
+ *         count++;
+ *         printf("Main: Set count to %lu\r\n", count);
+ *         cond.notify_all();
+ *         mutex.unlock();
+ *
+ *         wait(1.0);
+ *     }
+ *
+ *     mutex.lock();
+ *     // Change done and notify waiters of this
+ *     done = true;
+ *     printf("Main: Set done\r\n");
+ *     cond.notify_all();
+ *     mutex.unlock();
+ *
+ *     thread.join();
+ * }
+ * @endcode
+ */
+class ConditionVariable : private mbed::NonCopyable<ConditionVariable> {
+public:
+    /** Create and Initialize a ConditionVariable object */
+    ConditionVariable(Mutex &mutex);
+
+    /** Wait for a notification
+     *
+     * Wait until a notification occurs.
+     *
+     * @note - The thread calling this function must be the owner of the
+     * ConditionVariable's mutex and it must be locked once and only once
+     *
+     * Example:
+     * @code
+     * mutex.lock();
+     * while (!condition_met) {
+     *     timeout = cond.wait();
+     * }
+     * mutex.unlock();
+     * @endcode
+     */
+    void wait();
+
+    /** Wait for a notification or timeout
+     *
+     * @param   millisec  timeout value or 0 in case of no time-out.
+     * @return  true if woken from notification, false if timed out.
+     *
+     * @note - The thread calling this function must be the owner of the
+     * ConditionVariable's mutex and it must be locked once and only once
+     * @note - Spurious notifications can occur so the caller of this API should check the
+     *
+     * Example:
+     * @code
+     * mutex.lock();
+     * bool woke_from_notify = false;
+     * while (!condition_met && woke_from_notify) {
+     *     woke_from_notify = cond.wait_for(time_left);
```

Comment:
I still have mega-reservations about `wait_for` without `wait_until` - you didn't respond to my last diatribe, which is somewhere in an outdated version. (Find and read if you missed it).

This example totally glosses over how they're supposed to determine `time_left`.

I understand the issue about ticks versus milliseconds and wraps, but I don't think that problem not being resolved merits leaving a landmine in this API. I'd rather there was no relative timed wait in the condvar API at all, like pthreads and C11, so I don't want it being the only timed wait as a sort of interim measure, and hence used.

Can we defer it and try to pin down what would be required in core API to permit `wait_until`? I think it would be quite similar to this PR which replaced a single "relative milliseconds" API to a suite of "in/after/every/at" calls using ticks, with support macros: https://github.com/ARMmbed/sal-stack-nanostack-eventloop/pull/28

Or, could we do something inside ConditionVariable itself, pending a system-wide change? Some sort of auxiliary API giving access to an absolute end time for wait until? I think this would be a minimal as-abstract-as-possible definition:

    // Return an absolute time suitable to be passed to wait_until representing "now" plus the
    // specified number of milliseconds, rounded down - wait_until will return at or just before the specified
    // number of milliseconds from now.
    static abs_time_t get_absolute_time_by_ms(int32_t millisec);

    // Return an absolute time suitable to be passed to wait_until representing "now" plus the
    // specified number of milliseconds, rounded up - wait_until will return at or just after the specified
    // number of milliseconds from now.
    static abs_time_t get_absolute_time_after_ms(int32_t millisec);

    // example use:
    // returns true if my condition is met, false if it's not (in which case we must have timed out)
    bool wait_for_my_condition(int32_t millisec)
    {
         // short-cut just to avoid end time computation if already true
         if (my_condition()) {
             return true;
         }

        ConditionVariable::abs_time_t end_time = cond.get_absolute_time_after_ms(millisec);
        do {
              if (cond.wait_until(end_time) == osErrTimedOut) {
                   return my_condition();
        } while (!my_condition());

        return true;
    }

Reply:
Yep, I looked at your previous post. I would argue this is important enough to add in this PR, as ConditionVariable would not be usable in many use cases without  it. I updated the example to include how to safely use by updating the timeout with a timer.

I don't think the API of ConditionVariable should be extended as a workaround from a missing absolute rtos time support, as these functions would just be temporary and would add confusion in the long term with both APIs are available.

Is the update example enough for you to work with for now?  I have no problem with adding ```wait_until``` in the future, but I would argue it is out of the scope of this PR to add now as it would require a system-wide change.